File: bun-mqtt-backend/src/utils/timestamp.ts
/**
 * 时间戳单位智能识别工具。
 *
 * 背景：InfluxDB writeApi 使用纳秒精度，但不同厂商的传感器可能发送
 * 秒 / 毫秒 / 微秒 / 纳秒，如果不做识别会导致数据落到 1970 年附近。
 *
 * 判定策略（按数量级）：
 *   - 秒 (s)   ≈ 1.7e9    → 10 位数
 *   - 毫秒 (ms) ≈ 1.7e12   → 13 位数
 *   - 微秒 (μs) ≈ 1.7e15   → 16 位数
 *   - 纳秒 (ns) ≈ 1.7e18   → 19 位数
 *
 * 阈值取 1e11 / 1e14 / 1e17，可覆盖到公元 ~5138 年，绝对安全。
 */

export type TimestampUnit = "s" | "ms" | "us" | "ns" | "invalid";

export interface TimestampNormalizeResult {
  /** 归一化后的纳秒时间戳；无效输入时为 fallback（当前时间纳秒） */
  nanos: number;
  /** 识别到的原始单位 */
  unit: TimestampUnit;
  /** 是否使用了 fallback（原始输入无效） */
  usedFallback: boolean;
}

/** 仅识别单位，不做转换 */
export function detectTimestampUnit(ts: number): TimestampUnit {
  if (!Number.isFinite(ts) || ts <= 0) return "invalid";
  if (ts < 1e11) return "s";
  if (ts < 1e14) return "ms";
  if (ts < 1e17) return "us";
  return "ns";
}

/**
 * 把任意单位的时间戳归一化为纳秒。
 * 无效输入（NaN、负数、非有限数）会返回当前时间纳秒作为兜底。
 */
export function normalizeToNanoseconds(ts: number | undefined | null): TimestampNormalizeResult {
  const fallbackNanos = Date.now() * 1_000_000;

  if (ts === undefined || ts === null) {
    return { nanos: fallbackNanos, unit: "invalid", usedFallback: true };
  }

  const unit = detectTimestampUnit(ts);
  if (unit === "invalid") {
    return { nanos: fallbackNanos, unit: "invalid", usedFallback: true };
  }

  let nanos: number;
  switch (unit) {
    case "s":
      nanos = ts * 1_000_000_000;
      break;
    case "ms":
      nanos = ts * 1_000_000;
      break;
    case "us":
      nanos = ts * 1_000;
      break;
    case "ns":
      nanos = ts;
      break;
  }

  return { nanos, unit, usedFallback: false };
}

File: rust-mqtt-backend/src/timestamp.rs
//! 智能时间戳单位识别与纳秒归一化（对齐 go-mqtt-backend/internal/timestamp）。
//!
//! 阈值与倍率见 HANDOFF.md 第 2.6 节。

use std::time::{SystemTime, UNIX_EPOCH};

/// 时间戳单位（对齐 2.6）。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Unit {
    Second,
    Millisecond,
    Microsecond,
    Nanosecond,
    Invalid,
}

impl Unit {
    pub fn as_str(&self) -> &'static str {
        match self {
            Unit::Second => "s",
            Unit::Millisecond => "ms",
            Unit::Microsecond => "us",
            Unit::Nanosecond => "ns",
            Unit::Invalid => "invalid",
        }
    }
}

/// 归一化结果（对齐 Go NormalizeResult）。
#[derive(Debug, Clone)]
pub struct NormalizeResult {
    pub nanos: i64,
    pub unit: Unit,
    pub used_fallback: bool,
}

/// 当前系统时间纳秒（⚠️ 陷阱 5：as_nanos 是 u128，2262 年前转 i64 不溢出）。
fn now_nanos() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos() as i64)
        .unwrap_or(0)
}

/// 根据数量级判定时间戳单位：<1e11 秒；<1e14 毫秒；<1e17 微秒；>=1e17 纳秒。
pub fn detect_unit(ts: f64) -> Unit {
    if ts.is_nan() || ts.is_infinite() || ts <= 0.0 {
        return Unit::Invalid;
    }
    if ts < 1e11 {
        Unit::Second
    } else if ts < 1e14 {
        Unit::Millisecond
    } else if ts < 1e17 {
        Unit::Microsecond
    } else {
        Unit::Nanosecond
    }
}

/// 将任意单位的时间戳归一化为纳秒；None 或无效则用当前时间兜底。
pub fn normalize_to_nanos(ts: Option<f64>) -> NormalizeResult {
    let fallback_nanos = now_nanos();

    let ts = match ts {
        Some(v) => v,
        None => {
            return NormalizeResult {
                nanos: fallback_nanos,
                unit: Unit::Invalid,
                used_fallback: true,
            }
        }
    };

    let unit = detect_unit(ts);
    if unit == Unit::Invalid {
        return NormalizeResult {
            nanos: fallback_nanos,
            unit: Unit::Invalid,
            used_fallback: true,
        };
    }

    // ⚠️ 陷阱：先 `ts as i64` 再乘整数倍率，避免 f64 精度丢失。
    let nanos = match unit {
        Unit::Second => (ts as i64) * 1_000_000_000,
        Unit::Millisecond => (ts as i64) * 1_000_000,
        Unit::Microsecond => (ts as i64) * 1_000,
        Unit::Nanosecond => ts as i64,
        Unit::Invalid => unreachable!(),
    };

    NormalizeResult {
        nanos,
        unit,
        used_fallback: false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fp(v: f64) -> Option<f64> {
        Some(v)
    }

    // ---------- detect_unit（5） ----------

    #[test]
    fn detect_unit_second() {
        assert_eq!(Unit::Second, detect_unit(1_758_470_400.0));
        assert_eq!(Unit::Second, detect_unit(1.0));
        assert_eq!(Unit::Second, detect_unit(9.9e10));
    }

    #[test]
    fn detect_unit_millisecond() {
        assert_eq!(Unit::Millisecond, detect_unit(1_758_470_400_000.0));
        let now_ms = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as f64;
        assert_eq!(Unit::Millisecond, detect_unit(now_ms));
        assert_eq!(Unit::Millisecond, detect_unit(1e11));
    }

    #[test]
    fn detect_unit_microsecond() {
        assert_eq!(Unit::Microsecond, detect_unit(1_758_470_400_000_000.0));
        assert_eq!(Unit::Microsecond, detect_unit(1e14));
    }

    #[test]
    fn detect_unit_nanosecond() {
        assert_eq!(Unit::Nanosecond, detect_unit(1_758_470_400_000_000_000.0));
        assert_eq!(Unit::Nanosecond, detect_unit(1e17));
    }

    #[test]
    fn detect_unit_invalid() {
        assert_eq!(Unit::Invalid, detect_unit(0.0));
        assert_eq!(Unit::Invalid, detect_unit(-1.0));
        assert_eq!(Unit::Invalid, detect_unit(f64::NAN));
        assert_eq!(Unit::Invalid, detect_unit(f64::INFINITY));
    }

    // ---------- normalize_to_nanos（7） ----------

    #[test]
    fn normalize_second() {
        let r = normalize_to_nanos(fp(1_758_470_400.0));
        assert_eq!(Unit::Second, r.unit);
        assert!(!r.used_fallback);
        assert_eq!(1_758_470_400_000_000_000i64, r.nanos);
    }

    #[test]
    fn normalize_millisecond() {
        let r = normalize_to_nanos(fp(1_758_470_400_000.0));
        assert_eq!(Unit::Millisecond, r.unit);
        assert_eq!(1_758_470_400_000_000_000i64, r.nanos);
    }

    #[test]
    fn normalize_microsecond() {
        let r = normalize_to_nanos(fp(1_758_470_400_000_000.0));
        assert_eq!(Unit::Microsecond, r.unit);
        assert_eq!(1_758_470_400_000_000_000i64, r.nanos);
    }

    #[test]
    fn normalize_nanosecond() {
        let r = normalize_to_nanos(fp(1_758_470_400_000_000_000.0));
        assert_eq!(Unit::Nanosecond, r.unit);
        assert_eq!(1_758_470_400_000_000_000i64, r.nanos);
    }

    #[test]
    fn normalize_nil_fallback() {
        let before = now_nanos();
        let r = normalize_to_nanos(None);
        let after = now_nanos();
        assert!(r.used_fallback);
        assert_eq!(Unit::Invalid, r.unit);
        assert!(r.nanos >= before);
        assert!(r.nanos <= after);
    }

    #[test]
    fn normalize_negative_fallback() {
        let r = normalize_to_nanos(fp(-100.0));
        assert!(r.used_fallback);
    }

    #[test]
    fn normalize_all_units_equal() {
        let seconds = 1_758_470_400.0f64;
        let expected = 1_758_470_400_000_000_000i64;
        assert_eq!(expected, normalize_to_nanos(fp(seconds)).nanos);
        assert_eq!(expected, normalize_to_nanos(fp(seconds * 1e3)).nanos);
        assert_eq!(expected, normalize_to_nanos(fp(seconds * 1e6)).nanos);
        assert_eq!(expected, normalize_to_nanos(fp(seconds * 1e9)).nanos);
    }
}
